genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/FileSystemFetchingCacheService.java

```diff
@@ -0,0 +1,343 @@
+                    uriString,
+                    resourceCacheId
+                );
+
+                // Download the resource into the cache
+                final File cachedResourceVersionDownloadFile = getCacheResourceVersionDownloadFile(
+                    resourceCacheId,
+                    resourceLastModified
+                );
+                try (
+                    final InputStream in = resource.getInputStream();
+                    final OutputStream out = new FileOutputStream(cachedResourceVersionDownloadFile)
+                ) {
+                    FileCopyUtils.copy(in, out);
+                    Files.move(cachedResourceVersionDownloadFile, cachedResourceVersionDataFile);
+                }
+            } else {
+                log.info(
+                    "Cache hit: {} (id: {})",
+                    uriString,
+                    resourceCacheId
+                );
+            }
+
+            Files.copy(cachedResourceVersionDataFile, destinationFile);
+        }
+
+        //Clean up any older versions in a different thread
+        oldResourcesCleanUpService.submit(() -> {
+            try {
+                cleanUpOlderResourceVersions(resourceCacheId, resourceLastModified);
+            } catch (IOException e) {
+                log.error(
+                    "Error cleaning up old resource resourceCacheId - {}, resourceLastModified - {}",
+                    resourceCacheId,
+                    resourceLastModified,
+                    e
+                );
+            }
+        });
+    }
+
+    private FileChannel getLockFileChannel(
+        final String resourceCacheId,
+        final long lastModified
+    ) throws FileNotFoundException {
+        return new RandomAccessFile(
+            getCacheResourceVersionLockFile(
+                resourceCacheId,
+                lastModified
+            ),
+            "rws"
+        ).getChannel();
+    }
+
+    private FileChannel getLockFileChannel(final File resourceVersionDir) throws FileNotFoundException {
+        final String resourceCacheId = resourceVersionDir.getParentFile().getName();
+        final long resourceLastModifiedTimeStamp = getResourceLastModified(resourceVersionDir);
+
+        return getLockFileChannel(resourceCacheId, resourceLastModifiedTimeStamp);
+    }
+
+    @VisibleForTesting
+    String getResourceCacheId(final URI uri) {
+        return DigestUtils.md5DigestAsHex(uri.toASCIIString().getBytes(StandardCharsets.UTF_8));
+    }
+
+    /**
+     * Helper to extract the resource last modified timestamp from the resource
+     * version directory handle.
+     *
+     * @param resourceVersionDir handle to the resource version directory
+     * @return resource last modified timestamp
+     */
+    private long getResourceLastModified(final File resourceVersionDir) {
+        return Long.parseLong(resourceVersionDir.getName());
+    }
+
+    private void createDirectoryStructureIfNotExists(final File dir) {
```

Comment:
This actually just creates the parent folders for a given file, doesn't create any structure, I find the name a bit misleading.
Moreover it swallows `IO` and throws `Runtime` what's the rationale for this?

Reply:
I had copied this piece of code from `NaiveFetchingCache` constructor and moved it in a method. I honestly did not pay much attention to the conversion of exceptions. I think we can keep them as checked exceptions. For the naming - the initial directory structure needed for the Cache was being created here. The name looks fine to me as that's what it is doing - creating a dir structure if it does not exist.